include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -5845,16 +5848,16 @@ inline DEAL_II_ALWAYS_INLINE VectorizedArrayType
   AssertIndexRange(q_point, this->n_quadrature_points);
   Assert(this->jacobian != nullptr, ExcNotInitialized());
 
-  VectorizedArrayType divergence;
+  VectorizedArrayType        divergence;
+  const VectorizedArrayType *gradients = this->gradients_quad + q_point;
```

Comment:
I must admit I don't like this optimization: it is hard to read. I would prefer below `gradients[(d * dim + e) * nqp + q_point]`. The compiler should be able to reuse the computed indices.

Reply:
I actually started with what you suggest before but it turns out the compiler can't make the optimization because it is not equivalent as it is written now. The compiler can't know that `q_point + (d * dim + e) * nqp` does not overflow and wrap around the 32 bit integer limit, so it has to evaluate them inside the loop. This would have been avoided if we used signed integers instead of unsigned ones because there overflow is undefined, so the optimization should have been valid. I think the issue is avoided if I instead declare `npg` as `std::size_t`. Then, the expression `q_point + (d * dim + e) * nqp` will all be evaluated in 64 bit integer type and the compiler is back into business, because `ptr + offset` cannot overflow. I don't have time to try it now, but I will do so.

Reply:
@kronbichler Do you still want to investigate this? Or do you want to rebase and merge?

Reply:
I will investigate it soon I hope, but not this week.

Reply:
This is now ready for re-review. I replaced the offsets by `size_t` and at least in the scalar case (step-37, step-48, step-59) the compiler generates good code. I'm investigating step-67 but I think we can always improve something later on given the current state.